Fixes to dragzone scaling, doubleclick support, multiple drag zones V2

grbd opened the pull request:
I think I've got it figured now, originaly the click point was being scaled
but what needs to be scaled is the offsets or markers as to the edge of the drag zone instead
as the buttons will shrink or grow in size based on the dpi settings
I've also added support for multiple drag zones

I think in the case of https://github.com/chromelyapps/Chromely.Legacy/blob/master/src/Chromely.Core/Host/FramelessDragRegion.cs#L50
It's legacy code that I think is calculating the size of the drag area based on windows controls or some kind of native method

However I think the more typical use case for frameless is to use the controls within the UI framework
which will be different sizes based on the css or the UI in use (such as Bootstrap / Foundation / Quasar / Vuetify etc)
So I think the best approach is to just to give a sane default and let the end user customise it

## Double click Drag zone

There's now support for a double click event handler on the drag zone under FramelessOption
ideally this would restore / maximise the window when triggered as a default
but I don't think we have a OS independent way of doing that at the moment via IChromelyNativeHost, so I just leave it up to the end user to override / implement

## Testing

I've tried this under Windows 10 with multiple dpi settings and with EnableHighDpiSupport disabled and enabled and under a few other windows os's such as Win7, WinServer 2012, WinServer 2016
It all seems fine

I've been using this to test

  * https://github.com/Hecatron/Hecatron.Scraps/tree/master/Chromely.Examples/Chromely.Quasar1

mattkol commented:
Great job!

Please can you add a default implementation for DblClickCallbackFunc. We could restrict it to Windows for now. Most developers just want to get things working. Not everyone has that indept skills.  Besides our focus now is just for Windows on Frameless. We can than add enough notes for those who want to override.

I may look closely during the week.

Thanks a lot for your efforts on this.

grbd commented:
Okay no problem
I was planning on putting this in on a follow up pull request but what I've done is modify the IChromelyNativeHost interface
to add members for SetWindowState / GetWindowIsMaximized
I wasn't sure if doing it this way would be okay
but I think it makes it cross platform and allows me to call those functions via nativemethods more easily
If you think it should be done a different way then let me know

for now I've added empty implementations for mac / linux 
and just added real implementations for windows

I do have a seperate question on accessing nativemethods via the chromelyapp instance
but I'm going to file the details for that in a seperate issue since I think I know what the problem is

mattkol commented:
The main point of making IChromelyNativeHost  as light as possible is to allow for customization. But if commonly used functionalities can be identified I think we can add them. Problem is it is a lot more work to ensure they work cross platform. 

I personally do not have that much bandwidth now to work on both Linux and Mac so I guess it will be ok to add that to a TODO list.

mattkol commented:
Adding Minimize option - cool! I wonder how we missed that. 👍

mattkol commented:
So if I get it right, effectively click events should be handled in the [Child WndProc](https://github.com/Hecatron-Forks/Chromely/blob/f1d997d8964942280e52a81391ce2cb54d285fdb/src/Chromely/Native/WinAPI/WindowMessageInterceptor.cs#L151). 

Really smart!

mattkol commented:
I think we should add another nuget update for v77 with this. Please let me know if there is any improvement coming soon. We can do it between now and the weekend.

Thanks for all the nice catches and efforts on this.

mattkol commented:
There is something I have noticed, even before now. I want to see if you notice it too.
The first time I launch the app, the window is blank until I click on it or when I have any mouse movement. Is it something you have seen?

Also is it not appropriate that we capture all the click events in the drag area in the child wndproc?

grbd commented:
Hi,
I don't have any other improvements at the moment so feel free to put another version out

the next thing I might look at is allowing for a full screen mode without kiosk mode enabled if that's possible using the same function call for changing the window state.
I also found this which might be a starting point for gtk3 frameless mode although it needs a drag handler (I'm not sure if it's gtk3)
https://gist.github.com/jeandudey/02dc5146864cd0abc30f

I didn't notice the black screen thing you've mentioned
bit I did recently updated my nvidia drivers to 445.75, but it turns out it's been creating some graphics glitches that effects people differently
(in my case after coming back from sleep I'm noticing graphics corruption on firefox windows)
but that might be a separate issue

grbd commented:
One thing I have noticed is that when frameless, maximised ends up = full screen.
I'm not sure if that's fixable or not I think it's related to the window styles under GetWindowStyles
